pkg/gardenlet/controller/networkpolicy/add.go

```diff
@@ -0,0 +1,130 @@
+			v1beta1constants.GardenRole: v1beta1constants.GardenRoleShoot,
+		})
+	}
+	if r.Resolver == nil {
+		resolver, err := hostnameresolver.CreateForCluster(seedCluster.GetConfig(), mgr.GetLogger())
+		if err != nil {
+			return fmt.Errorf("failed to get hostnameresolver: %w", err)
+		}
+		r.Resolver = resolver
+	}
+
+	return builder.
+		ControllerManagedBy(mgr).
+		Named(ControllerName).
+		For(&corev1.Namespace{}).
+		Watches(
+			source.NewKindWithCache(&corev1.Endpoints{}, seedCluster.GetCache()),
+			mapper.EnqueueRequestsFrom(mapper.MapFunc(r.MapToNamespaces), mapper.UpdateWithNew, mgr.GetLogger()),
+			builder.WithPredicates(r.IsKubernetesEndpoint()),
+		).
+		Watches(
+			source.NewKindWithCache(&networkingv1.NetworkPolicy{}, seedCluster.GetCache()),
+			mapper.EnqueueRequestsFrom(mapper.MapFunc(r.MapObjectToNamespace), mapper.UpdateWithNew, mgr.GetLogger()),
+			builder.WithPredicates(r.IsAllowToSeedApiserverNetworkPolicy()),
+		).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: *r.Config.ConcurrentSyncs,
+			RecoverPanic:            true,
+			RateLimiter:             workqueue.DefaultControllerRateLimiter(),
+		}).
+		Complete(r)
+}
+
+// MapToNamespaces is a mapper function which returns requests for all shoot namespaces + garden namespace + istio-system namespace.
+func (r *Reconciler) MapToNamespaces(ctx context.Context, log logr.Logger, _ client.Reader, _ client.Object) []reconcile.Request {
+	namespaces := &corev1.NamespaceList{}
+	if err := r.SeedClient.List(ctx, namespaces, &client.ListOptions{
+		LabelSelector: r.ShootNamespaceSelector,
+	}); err != nil {
+		log.Error(err, "Unable to list Shoot namespace for updating NetworkPolicy", "networkPolicyName", helper.AllowToSeedAPIServer)
+		return []reconcile.Request{}
+	}
+
+	requests := []reconcile.Request{}
+	for _, namespace := range namespaces.Items {
+		requests = append(requests, reconcile.Request{NamespacedName: client.ObjectKeyFromObject(&namespace)})
+	}
+	requests = append(requests, reconcile.Request{NamespacedName: types.NamespacedName{Name: v1beta1constants.GardenNamespace}})
+	requests = append(requests, reconcile.Request{NamespacedName: types.NamespacedName{Name: v1beta1constants.IstioSystemNamespace}})
+
+	return requests
+}
+
+// MapObjectToNamespace is a mapper function which maps an object to its namespace.
+func (r *Reconciler) MapObjectToNamespace(_ context.Context, _ logr.Logger, _ client.Reader, obj client.Object) []reconcile.Request {
+	return []reconcile.Request{{NamespacedName: types.NamespacedName{Name: obj.GetNamespace()}}}
+}
+
+// IsAllowToSeedApiserverNetworkPolicy returns a predicate which evaluates if the object the allow-to-seed-apiserver network policy.
+func (r *Reconciler) IsAllowToSeedApiserverNetworkPolicy() predicate.Predicate {
+	return predicate.NewPredicateFuncs(func(obj client.Object) bool {
+		policy, ok := obj.(*networkingv1.NetworkPolicy)
+		if !ok {
+			return false
+		}
+		return policy.Name == helper.AllowToSeedAPIServer
```

Comment:
No need to cast to `*networkingv1.NetworkPolicy`. You can use `obj.GetName()` instead.